src/crawlee/_utils/persistent_object.py

```diff
@@ -0,0 +1,152 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Generic, TypeVar
+
+from pydantic import BaseModel
+
+from crawlee import service_locator
+from crawlee.events._types import Event, EventPersistStateData
+from crawlee.storages._key_value_store import KeyValueStore
+
+if TYPE_CHECKING:
+    import logging
+
+TStateModel = TypeVar('TStateModel', bound=BaseModel)
+
+
+class PersistentObject(Generic[TStateModel]):
+    """A class for managing persistent state using a Pydantic model.
+
+    This class facilitates state persistence to a `KeyValueStore`, allowing data to be saved and retrieved
+    across migrations or restarts. It manages the loading, saving, and resetting of state data,
+    with optional persistence capabilities.
+
+    The state is represented by a Pydantic model that can be serialized to and deserialized from JSON.
+    The class automatically hooks into the event system to persist state when needed.
+
+    Type Parameters:
+        TStateModel: A Pydantic BaseModel type that defines the structure of the state data.
+                     Typically, it should be inferred from the `default_state` constructor parameter.
+    """
+
+    def __init__(
+        self,
+        *,
+        default_state: TStateModel,
+        persist_state_key: str,
+        persistence_enabled: bool = False,
+        persist_state_kvs_name: str | None = None,
+        persist_state_kvs_id: str | None = None,
+        logger: logging.Logger,
+    ) -> None:
+        """Initialize a new persistent object.
+
+        Args:
+            default_state: The default state model instance to use when no persisted state is found.
+                           A deep copy is made each time the state is used.
+            persist_state_key: The key under which the state is stored in the KeyValueStore
+            persistence_enabled: Flag to enable or disable state persistence
+            persist_state_kvs_name: The name of the KeyValueStore to use for persistence.
+                                    If neither a name nor and id are supplied, the default store will be used.
+            persist_state_kvs_id: The identifier of the KeyValueStore to use for persistence.
+                                    If neither a name nor and id are supplied, the default store will be used.
+            logger: A logger instance for logging operations related to state persistence
+        """
+        self._default_state = default_state
+        self._state_type: type[TStateModel] = self._default_state.__class__
+        self._state: TStateModel | None = None
+        self._persistence_enabled = persistence_enabled
+        self._persist_state_key = persist_state_key
+        self._persist_state_kvs_name = persist_state_kvs_name
+        self._persist_state_kvs_id = persist_state_kvs_id
+        self._key_value_store: KeyValueStore | None = None
+        self._log = logger
+
+    async def initialize(self) -> TStateModel:
```

Comment:
When I see `initialize` and `teardown`, I have to ask - shouldn't this be a context manager?

Reply:
I also considered this, but nothing tipped the scales for me. Especially the usage in `KeyValueStore` would feel dirty if we called `__aenter__` there directly.